File: priv/alphonse/lib/kernel.js
import Atom from './atom';
import Tuple from './tuple';

let Kernel = {
  __MODULE_: [Atom('Kernel')],

  tl: function(list){
    return list.slice(1);
  },

  hd: function(list){
    return list.slice(0,1)[0];
  },

  is_nil: function(x){
    return x == null;
  },

  is_atom: function(x){
    return x instanceof Atom;
  },

  is_binary: function (x){
    return typeof(x) === 'string' || x instanceof String;
  },

  is_boolean: function (x){
    return typeof(x) === 'boolean' || x instanceof Boolean; 
  },

  is_function: function(x, arity = -1){
    return x instanceof Function;
  },

  // from: http://stackoverflow.com/a/3885844
  is_float: function(x){
    return x === +x && x !== (x|0);
  },

  is_integer: function(x){
    return x === +x && x === (x|0);
  },

  is_list: function(x){
    return x instanceof Array;
  },

  is_map: function(x){
    return x instanceof Object;
  },

  is_number: function(x){
    return Kernel.is_integer(x) || Kernel.is_float(x);
  },

  is_tuple: function(x){
    return x instanceof Tuple;
  },

  length: function(x){
    return x.length;
  },

  is_pid: function(x){
    return false;
  },

  is_port: function(x){
    
  },

  is_reference: function(x){

  },

  is_bitstring: function(x){
    return Kernel.is_binary(x) || x instanceof BitString;
  },

  _in: function(left, right){
    return Enum.member(right, left);
  },

  abs: function(number){
    return Math.abs(number);
  },

  round: function(number){
    return Math.round(number);
  },

  elem: function(tuple, index){
    return tuple['_' + index];
  },

  rem: function(left, right){
    return left % right;
  },

  div: function(left, right){
    return left / right;
  },

  and: function(left, right){
    return left && right;
  },

  or: function(left, right){
    return left || right;
  },

  not: function(arg){
    return !arg;
  },

  apply: function(module, fun, args){
    if(arguments.length === 3){
      return module[fun].apply(null, args);
    }else{
      return module.apply(null, fun);  
    }
  },

  to_string: function(arg){
    return arg.toString();
  },

  match: function(pattern, expr){
    if(pattern === undefined){
      return true;
    }

    if(Kernel.is_nil(expr) || Kernel.is_number(expr) || Kernel.is_binary(expr) || Kernel.is_boolean(expr)){
      return pattern === expr;
    }else if(Kernel.is_atom(expr)){
      return Kernel.is_atom(pattern) && pattern.value === expr.value;
    }else if(Kernel.is_tuple(expr)){
      return Kernel.is_tuple(pattern) && Kernel.match(pattern.value, expr.value);
    }else if(Kernel.is_list(expr)){     
      if(Kernel.length(pattern) != Kernel.length(expr)){
        return false;
      }

      for (let i=0; i <= pattern.length; i++) {
        if(Kernel.match(pattern[i], expr[i]) === false){
          return false;
        }
      }

      return true;
    }else if(Kernel.is_map(expr)){
      if(!Kernel.is_map(pattern)){
        return false;
      }

      for(let key in pattern){
        if(!(key in expr)){
          return false;
        }

        if(Kernel.match(pattern[key], expr[key]) == false){
          return false;
        }
      }

      return true;
    }
  }
}

export default Kernel;

File: priv/std_lib/kernel.ex
defmodule ElixirScript.Kernel do
  @moduledoc false
  import Kernel, only: [defmodule: 2, def: 1, def: 2, defp: 2,
                        defmacro: 1, defmacro: 2, defmacrop: 2, ||: 2, !: 1,
                        ++: 2, in: 2, &&: 2, ===: 2, @: 1, sigil_r: 2]
  require JS


  defmacro if(condition, clauses) do
    build_if(condition, clauses)
  end

  defp build_if(condition, do: do_clause) do
    build_if(condition, do: do_clause, else: nil)
  end

  defp build_if(condition, do: do_clause, else: else_clause) do
    quote do
      case unquote(condition) do
        x when x in [false, nil] ->
          unquote(else_clause)
        _ ->
          unquote(do_clause)
      end
    end
  end

  defmacro unless(condition, clauses) do
    build_unless(condition, clauses)
  end

  defp build_unless(condition, do: do_clause) do
    build_unless(condition, do: do_clause, else: nil)
  end

  defp build_unless(condition, do: do_clause, else: else_clause) do
    quote do
      if(unquote(condition), do: unquote(else_clause), else: unquote(do_clause))
    end
  end

  def abs(number) do
    JS.Math.abs(number)
  end

  def apply(fun, args) do
    fun.apply(fun, args)
  end

  def apply(module, fun, args) do
    module[Atom.to_string(fun)].apply(module[Atom.to_string(fun)], args)
  end

  def binary_part(binary, start, len) do
    binary.substring(start, len)
  end

  def hd(list) do
    list[0]
  end

  def tl(list) do
    list.slice(1)
  end

  def is_atom(term) do
    JS.typeof(term) === "symbol"
  end

  def is_binary(term) do
    JS.typeof(term) === "string"
  end

  def is_bitstring(term) do
    is_binary(term) || JS.instanceof(term, Bootstrap.Core.BitString)
  end

  def is_boolean(term) do
    JS.typeof(term) === "boolean" || JS.instanceof(term, Boolean)
  end

  def is_float(term) do
    is_number(term) && !JS.Number.isInteger(term)
  end

  def is_function(term) do
    is_function(term, 0)
  end

  def is_function(term, _) do
    JS.typeof(term) === "function" || JS.instanceof(term, Function)
  end

  def is_integer(term) do
    JS.Number.isInteger(term)
  end

  def is_list(term) do
    JS.Array.isArray(term)
  end

  def is_number(term) do
    JS.typeof(term) === "number" || JS.instanceof(term, Number)
  end

  def is_pid(term) do
    JS.instanceof(term, Bootstrap.Core.PID)
  end

  def is_tuple(term) do
    JS.instanceof(term, Bootstrap.Core.Tuple)
  end

  def is_map(term) do
    JS.typeof(term) === "object" || JS.instanceof(term, Object)
  end

  def is_port(_) do
    false
  end

  def is_reference(_) do
    false
  end

  def length(term) do
    term.length
  end

  def map_size(term) do
    JS.Object.keys(term).length
  end

  def max(first, second) do
    JS.Math.max(first, second)
  end

  def min(first, second) do
    JS.Math.min(first, second)
  end

  def round(number) do
    JS.Math.round(number)
  end

  def trunc(number) do
    JS.Math.floor(number)
  end

  def tuple_size(tuple) do
    tuple.count()
  end

  def elem(tuple, index) do
    tuple.get(index)
  end

  def is_nil(term) do
    term === nil
  end

  defmacro sigil_r({:<<>>, _meta, [string]}, options) do
    str_options = List.to_string(options)
    quote do
      Regex.compile!(unquote(string), unquote(str_options))
    end
  end

  defmacro match?(pattern, expr) do
    quote do
      case unquote(expr) do
        unquote(pattern) ->
          true
        _ ->
          false
      end
    end
  end

  defmacro to_string(arg) when Kernel.is_binary(arg) do
    arg
  end

  defmacro to_string(arg) do
    quote do
      String.Chars.to_string(unquote(arg))
    end
  end

  defmacro left |> {fun, context, params} do
    {fun, context, [left] ++ params }
  end

  defmacro left in right do
    quote do
      ElixirScript.Bootstrap.Functions.contains(unquote(left), unquote(right))
    end
  end

  defmacro first .. last do
    quote do
      %ElixirScript.Range{ first: unquote(first), last: unquote(last) }
    end
  end
end
